lib/nx_gzlib.c

```diff
@@ -0,0 +1,295 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <errno.h>
+#include "nx_zlib.h"
+
+#define DEF_MEM_LEVEL 8
+#define MAX_LEN 4096
+
+gzFile nx_gzopen_(const char* path, int fd, const char *mode);
+
+typedef struct gz_state {
+	gzFile gz;
+	int fd;
+	int err;
+	z_stream strm;
+	bool deflate;
+} *gz_statep;
+
+int nx_gzclose(gzFile file)
+{
+	gz_statep state = (gz_statep) file;
+	z_streamp stream = &(state->strm);
+	unsigned char* next_out = malloc(sizeof(char)*MAX_LEN);
+	int ret;
+
+	if(file == NULL){
+		errno = EINVAL;
+		return Z_STREAM_ERROR;
+	}
+
+	if (state->deflate) {
+		/* Finish the stream.  */
+		stream->next_in = Z_NULL;
+		stream->avail_in = 0;
+		do {
+			stream->next_out = next_out;
+			stream->avail_out = MAX_LEN;
+
+			nx_deflate(stream, Z_FINISH);
+			write(state->fd, next_out, MAX_LEN - stream->avail_out);
+		} while (stream->avail_out == 0);
+
+		ret = nx_deflateEnd(stream);
+	} else
+		ret = nx_inflateEnd(stream);
+
+	free(next_out);
+	free(state);
+
+	return ret;
+}
+
+gzFile nx_gzdopen(int fd, const char *mode)
+{
+	return nx_gzopen_(NULL, fd, mode);
+}
+
+gzFile nx_gzopen(const char* path, const char *mode)
+{
+	return nx_gzopen_(path, 0, mode);
+}
+
+gzFile nx_gzopen_(const char* path, int fd, const char *mode)
+{
+	gz_statep state;
+	gzFile file;
+	int err, strategy = Z_DEFAULT_STRATEGY, level = Z_DEFAULT_COMPRESSION;
+	char* digit;
+	void* stream;
+
+	state = malloc(sizeof(struct gz_state));
+	memset(state, 0, sizeof(struct gz_state));
+
+	if (path == NULL){
+		if(fdopen(fd, mode) != NULL)
+			state->fd = fd;
+		else
+			state->fd = -1;
+	} else {
+		FILE* fp = fopen(path, mode);
+		state->fd = fileno(fp);
+	}
+
+	if (state->fd == -1){
+		free(state);
+		return NULL;
+	}
+
+	stream = &(state->strm);
+	memset(stream, 0, sizeof(z_stream));
+	if (strchr(mode, 'w')){
+		if (strchr(mode, 'h'))
+			strategy = Z_HUFFMAN_ONLY;
+		else if (strchr(mode, 'f'))
+			strategy = Z_FILTERED;
+		else if (strchr(mode, 'R'))
+			strategy = Z_RLE;
+
+		digit = strpbrk(mode, "0123456789");
+		if (digit != NULL)
+			level = atoi(digit);
+
+		err = nx_deflateInit2(stream, level, Z_DEFLATED, 31,
+					DEF_MEM_LEVEL, strategy);
+		state->deflate = true;
+	} else {
+		err = nx_inflateInit(stream);
+		state->deflate = false;
+	}
+	if (err != Z_OK){
+		free(state);
+		if (err == Z_STREAM_ERROR)
+			errno = EINVAL;
+		return NULL;
+	}
+
+	file = (gzFile) state;
+	return file;
+}
+
+int nx_gzread(gzFile file, void *buf, unsigned len)
+{
+	gz_statep state = (gz_statep) file;
+	z_streamp stream = &(state->strm);
+	int ret, err;
+	const uInt max = len > 10? 10 : 1;
```

Comment:
Is this correct? Why are you using such low values?

Reply:
Yes, that's one of the points I raised on https://github.com/libnxz/power-gzip/pull/138#issuecomment-1077786216